orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -300,6 +522,130 @@ Status TransformGraphForPipeline(
     backward_waited_event_name,
     forward_recorded_output_name,
     backward_waited_output_name));
+  
+  // Different stages have different patterns of Send & Recv.
+  // For different patterns, we add different WaitEvent and Record.
+  //
+  // After applying all transformations above, the current patterns
+  // are listed below.
+  //
+  // 1. First stage:
+  //  Wait -----------------> FW -> --------> Send -> Record ->
+  //  Wait -> Recv ---------> BW -> ----------------> Record
+  // 2. Middle stage:
+  //  Wait -> Recv -> Wait -> FW -> Record -> Send -> Record ->
+  //  Wait -> Recv -> Wait -> BW -> Record -> Send -> Record
+  // 3. Last stage:
+  //  Wait -> Recv ---------> FW ----------------------------->
+  //  ----------------------> BW -----------> Send -> Record
+  //
+  // After applying all transformations below, we will have
+  // the following patterns.
+  //
+  // 1. First stage:
+  //  Wait ---------> Wait -> FW -> Record -> Send -> Record ->
+  //  Wait -> Recv -> Wait -> BW -> Record ---------> Record
+  // 2. Middle stage:
+  //  Wait -> Recv -> Wait -> FW -> Record -> Send -> Record ->
+  //  Wait -> Recv -> Wait -> BW -> Record -> Send -> Record
+  // 3. Last stage:
+  //  Wait -> Recv -> Wait -> FW ----------------------------->
+  //  ----------------------> BW -> Record -> Send -> Record
+  const bool is_first_stage = !recv_fw && send_fw && recv_bw && !send_bw;
+  const bool is_middle_stage = recv_fw && send_fw && recv_bw && send_bw;
+  const bool is_last_stage = recv_fw && !send_fw && !recv_bw && send_bw;
```

Comment:
maybe add a check to ensure one and only one of these bools is true? Otherwise we would add bogus wait/record later.

Reply:
Sure. Thanks.